### PR TITLE
feat(rome_formatter): typescript support, part 2

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Part of #2069 

This PR formats:
- `extends` clause and the list of types inside `interface`
- tuple types
- array type
- it adds a new parameter to the function `format_separated`. In TS there are few lists that don't allow a trailing separator at the end of those lists, this new parameter tells the function if that trailing separator is allowed at the end of the list
- global declaration
- declare function declaration
- declaration statement
- non null expression
- module block
- as expression
- type assertion expression
- `format_separated` now accepts a `&L` instead of `L`, avoiding the cloning

## Test Plan

New tests added

### CODE REVIEW

File: crates/rome_formatter/src/ts/declarations/declare_function_declaration.rs [hunk 1]
```diff
@@ -1,7 +1,42 @@
...
+        })?;
+
+        let function_token = function_token.format(formatter)?;
+        let id = id.format(formatter)?;
+        let type_parameters = type_parameters.format_or_empty(formatter)?;
+        let parameters = parameters.format(formatter)?;
+        let return_type_annotation = return_type_annotation.format_or_empty(formatter)?;
+        let semicolon_token = semicolon_token.format_or(formatter, || token(";"))?;
+
+        Ok(format_elements![
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/ts/declarations/declare_function_declaration.rs:31]: Makes sense!

File: crates/rome_formatter/src/formatter.rs [hunk 2]
```diff
@@ -29,6 +29,18 @@ pub struct Formatter {
     pub(super) printed_tokens: RefCell<HashSet<SyntaxToken>>,
 }
 
+#[derive(Debug)]
+pub enum TrailingSeparator {
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/formatter.rs:33]: What's an example of those lists?

File: crates/rome_formatter/src/ts/declarations/declare_function_declaration.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/ts/declarations/declare_function_declaration.rs:32]: Done

File: crates/rome_formatter/src/formatter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leops) [reply crates/rome_formatter/src/formatter.rs:33]: In ES5 mode Prettier does not insert trailing commas in arguments / parameters lists

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/formatter.rs:33]: Considering this, I wonder if we should have some "internals" inside `Formatter`. I mean, we have `FormatOptions` which come from the user, but I think we should track these kind of information that depends from the language version. Ideally something like `FormatSettings`:
```rs
struct FormatSettings {
	trailing_separator: TrailingSeparator
}
```
 and then we could have a `From<Syntax> for FormatSettings` so we can deduce correctly the info we need. And yes, we could add a third option to `TrailingSeparator`. What do you think?

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_formatter/src/formatter.rs:33]: Makes sense, I would leave this for another PR in case we decided to have add this new setting

**Pull request merged**